mobile_verifier/src/coverage.rs

```diff
@@ -0,0 +1,466 @@
+use tokio::sync::mpsc::Receiver;
+use uuid::Uuid;
+
+#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Type)]
+#[sqlx(type_name = "signal_level")]
+#[sqlx(rename_all = "lowercase")]
+pub enum SignalLevel {
+    No,
+    Low,
+    Medium,
+    High,
+}
+
+impl From<SignalLevelProto> for SignalLevel {
+    fn from(level: SignalLevelProto) -> Self {
+        match level {
+            SignalLevelProto::High => Self::High,
+            SignalLevelProto::Medium => Self::Medium,
+            SignalLevelProto::Low => Self::Low,
+            SignalLevelProto::No => Self::Low,
+        }
+    }
+}
+
+pub struct CoverageDaemon {
+    pool: Pool<Postgres>,
+    auth_client: AuthorizationClient,
+    coverage_objs: Receiver<FileInfoStream<CoverageObjectIngestReport>>,
+    file_sink: FileSinkClient,
+}
+
+impl CoverageDaemon {
+    pub fn new(
+        pool: Pool<Postgres>,
+        auth_client: AuthorizationClient,
+        coverage_objs: Receiver<FileInfoStream<CoverageObjectIngestReport>>,
+        file_sink: FileSinkClient,
+    ) -> Self {
+        Self {
+            pool,
+            auth_client,
+            coverage_objs,
+            file_sink,
+        }
+    }
+
+    pub async fn run(mut self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = shutdown.clone() => {
+                        tracing::info!("CoverageDaemon shutting down");
+                        break;
+                    }
+                    Some(file) = self.coverage_objs.recv() => self.process_file(file).await?,
+                }
+            }
+
+            Ok(())
+        })
+        .map_err(anyhow::Error::from)
+        .and_then(|result| async move { result })
+        .await
+    }
+
+    async fn process_file(
+        &mut self,
+        file: FileInfoStream<CoverageObjectIngestReport>,
+    ) -> anyhow::Result<()> {
+        tracing::info!("Processing coverage object file {}", file.file_info.key);
+
+        let mut transaction = self.pool.begin().await?;
+        let mut reports = file.into_stream(&mut transaction).await?;
+
+        while let Some(CoverageObjectIngestReport { report, .. }) = reports.next().await {
+            let CoverageObject {
+                pub_key,
+                uuid,
+                cbsd_id,
+                coverage_claim_time,
+                indoor,
+                coverage,
+            } = report;
+
+            // TODO: Validate pub key is authorized
+            let validity = if self
+                .auth_client
+                .verify_authorized_key(&pub_key, NetworkKeyRole::MobilePcs)
+                .await?
+            {
+                CoverageObjectValidity::Valid
+            } else {
+                CoverageObjectValidity::InvalidPubKey
+            };
+
+            self.file_sink
+                .write(
+                    proto::CoverageObject {
```

Comment:
Should we be creating proto struct directly ?  i understood that the convention was to create an internal struct and then perform an into/try_into to get to the proto.  

If we have properly defined filestore conversions it ensures there are no mishaps with 
creating proto struct manually ( ie getting a timestamp precision incorrect )

We should clarify amongst ourselves if this convention is valid, as i see both approaches in use in different parts of the code.